### PR TITLE
[backport] Flag Deprecation (13.0)

### PR DESCRIPTION
**Pull request by ajm188**
This back ports #9733 to release-13.0.

### CODE REVIEW

File: go/internal/flag/flag.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Package flag is an internal package to allow us to gracefully transition
```
Comment (deepthi) [inline go/internal/flag/flag.go:1]: In theory we need a copyright section here (though we are in a bit of flux around what that will look like).

**Review by deepthi — APPROVED** [review]
Approving provisionally (assuming no real test failures).

**Comment by deepthi** [comment]
Second thoughts:
* What happens to the help output? Right now we print it out with a single `-` in front of each flag
* I see this in the main PR
```
W0217 09:40:27.230552   62398 flag.go:43] Detected a positional argument beginning with a dash; This will be treated as a flag argument in the next version of Vitess. Please update your invocation to include a "--" before to continue treating --root as a positional argument.
W0217 09:40:27.230567   62398 flag.go:43] Detected a positional argument beginning with a dash; This will be treated as a flag argument in the next version of Vitess. Please update your invocation to include a "--" before to continue treating --server-address as a positional argument.
```
That doesn't seem right. The command invocation is:
```
vtctl $TOPOLOGY_FLAGS VtctldCommand AddCellInfo \
  --root /vitess/$cell \
  --server-address "${ETCD_SERVER}" \
  $cell
```
which is using the flags correctly.

**Comment by ajm188** [comment]
> which is using the flags correctly.

it's not. the correct usage there is to defer those double-dashed arguments to the subcommand, so the correct way to do this with `pflag` is _any of_:
* `vtctl $TOPOLOGY_FLAGS VtctldCommand AddCellInfo -- --server-address ...` OR
* `vtctl $TOPOLOGY_FLAGS VtctldCommand -- AddCellInfo --server-address ...` OR
* `vtctl $TOPOLOGY_FLAGS -- VtctldCommand AddCellInfo --server-address ...`

**Comment by deepthi** [comment]
> > which is using the flags correctly.
> 
> it's not. the correct usage there is to defer those double-dashed arguments to the subcommand, so the correct way to do this with `pflag` is _any of_:
> 
> * `vtctl $TOPOLOGY_FLAGS VtctldCommand AddCellInfo -- --server-address ...` OR
> * `vtctl $TOPOLOGY_FLAGS VtctldCommand -- AddCellInfo --server-address ...` OR
> * `vtctl $TOPOLOGY_FLAGS -- VtctldCommand AddCellInfo --server-address ...`

Ah I missed that these go to the `AddCellInfo`. So that part is good.

**Comment by deepthi** [comment]
OK, I went back and looked at the sample output. This also looks correct.
```
W0217 09:40:27.230091   62398 flag.go:30] Use of single-dash long flags is deprecated and will be removed in the next version of Vitess. Please use --topo_global_root instead
W0217 09:40:27.230545   62398 flag.go:30] Use of single-dash long flags is deprecated and will be removed in the next version of Vitess. Please use --topo_global_server_address instead
W0217 09:40:27.230549   62398 flag.go:30] Use of single-dash long flags is deprecated and will be removed in the next version of Vitess. Please use --topo_implementation instead
```

That leaves only the question about the help. Do we print out help text with `-` or `--` with this change?

**Comment by ajm188** [comment]
labeling this with DoNotMerge as i've pushed up more changes to the original branch, and i'm skeptical i will get this done in time for the 13.0 release 😞

**Comment by ajm188** [comment]
> That leaves only the question about the help. Do we print out help text with `-` or `--` with this change?

I figured out how to work with the flag internals, and we are now printing out `--` with this change (in the upstream version, I'm working on updating the backport branch now)

**Comment by ajm188** [comment]
`v13` has already been created, so i'm closing this and will get it out for `v14`

**Pull request closed without merging**